core-plugins/src/main/java/co/cask/hydrator/plugin/batch/file/AbstractFileMetadata.java

```diff
@@ -160,11 +233,78 @@ public short getPermission() {
     return permission;
   }
 
+  public String getHostURI() {
+    return hostURI;
+  }
+
+  public String getFilesystem() {
+    return filesystem;
+  }
+
+  /**
+   * Converts to StructuredRecord
+   */
+  public StructuredRecord toRecord() {
+
+    // initialize output schema
+    Schema outputSchema;
+    List<Schema.Field> fieldList = new ArrayList<>(DEFAULT_SCHEMA.getFields());
+    fieldList.addAll(getCredentialSchema().getFields());
+    outputSchema = Schema.recordOf("metadata", fieldList);
+
+    StructuredRecord.Builder outputBuilder = StructuredRecord.builder(outputSchema)
+      .set(FILE_NAME, fileName)
+      .set(FULL_PATH, fullPath)
+      .set(FILE_SIZE, fileSize)
+      .set(MODIFICATION_TIME, modificationTime)
+      .set(GROUP, group)
+      .set(OWNER, owner)
+      .set(IS_FOLDER, isFolder)
+      .set(BASE_PATH, basePath)
+      .set(PERMISSION, permission)
+      .set(HOST_URI, hostURI);
+    addCredentialsToBuilder(outputBuilder);
+
+    return outputBuilder.build();
+  }
+
+  @Override
+  public int compareTo(Object o) {
+    return Long.compare(fileSize, ((AbstractFileMetadata) o).getFileSize());
+  }
+
+  @Override
+  public void write(DataOutput dataOutput) throws IOException {
+    dataOutput.writeUTF(getFileName());
+    dataOutput.writeUTF(getFullPath());
+    dataOutput.writeLong(getModificationTime());
+    dataOutput.writeUTF(getGroup());
+    dataOutput.writeUTF(getOwner());
+    dataOutput.writeLong(getFileSize());
+    dataOutput.writeBoolean(isFolder());
+    dataOutput.writeUTF(getBasePath());
+    dataOutput.writeShort(getPermission());
+    dataOutput.writeUTF(getFilesystem());
+    dataOutput.writeUTF(getHostURI());
+  }
+
   /**
-   * extend from this class to create a credential class for each filesystem
+   * Don't use this to deserialize AbstractFileMetadata. Use the constructor
+   * that takes DataInput as an input instead.
+   * @param dataInput
+   * @throws IOException
    */
-  public abstract static class Credentials {
-    public String databaseType;
-    public static final String DATA_BASE_TYPE = "databaseType";
+  @Override
+  @Deprecated
+  public void readFields(DataInput dataInput) throws IOException {
   }
+
+  /**
+   * @return Credential schema for different filesystems
+   */
+  protected abstract Schema getCredentialSchema();
+
+  protected abstract void addCredentialsToBuilder(StructuredRecord.Builder builder);
+
+  protected abstract String getFSName();
```

Comment:
where/how is this used? Is there a reason this is used instead of the scheme for the filesystem uri? (hdfs, s3a, s3n, ftp, etc.)